doc/internal/consensus/Consensus-Protocol.md

```diff
@@ -0,0 +1,127 @@
+# The consensus
+
+Before talking about consensus, we need to discuss some logistics based on how the systems can co-exist.
+
+* Communication: Distributed systems need a method to communicate between each other. Remote Procedure Calls is the mechanism using which a standalone server can talk to another. The existing `network` layer of the database will handle all the communication between servers. 
+* Security: Access control mechanisms need to be in place to decide on access to functions in the servers based on their state (leader, follower, candidate)
+* Routing to leader: One of the issues with a varying leader is for the clients to know which IP address to contact for the service. We can solve this problem by advertising any/all IPs of the cluster and simply forward this request to the current leader; OR have a proxy that can forward the request to the current leader wheneve the requests come in. (Section client interaction of post has another approach which works too)
+* The servers will be implemented in the `interal/node` folders which will import the raft API and perform their functions.
+
+Maintaining consensus is one of the major parts of a distributed system. To know to have achieved a stable system, we need the following two parts of implementation.
+
+## The Raft protocol
+
+The raft protocol will be implemented in `internal/raft` and will implement APIs that each node can call.
+
+Raft is an algorithm to handle replicated log, and we maintain the "log" of the SQL stmts applied on a DB and have a completely replicated cluster.
+
+#### General Implementation rules:
+* All RPC calls are done parallely to obtain the best performance.
+* Request retries are done in case of network failures.
+* Raft does not assume network preserves ordering of the packets.
+
+A raft server may be in any of the 3 states; leader, follower or candidate. All requests are serviced through the leader and it then decides how and if the logs must be replicated in the follower machines. The raft protocol has 3 almost independent modules:
+1. Leader Election
+2. Log Replication
+3. Safety
+
+A detailed description of all the modules follow:
+
+### Leader Election
+
+#### Spec
+* Startup: All servers start in the follower state and begin by requesting votes to be elected as a leader.
+* Pre-election: The server increments its `currentTerm` by one, changes to `candidate` state and sends out `RequestVotes` RPC parallely to all the peers. 
+* Vote condition: FCFS basis. If there was no request to the server, it votes for itself (read 3.6 and clear out when to vote for itself)
```

Comment:
3.6 where? Maybe add a link if possible

Reply:
That was a note for myself while reading ^^'
Removed.